### PR TITLE
pod resources per composition group

### PR DESCRIPTION
**Pull request by nonsense**
Addresses https://github.com/ipfs/testground/issues/789

TODO:
- [ ] ~check memory as well, not just CPU~

### CODE REVIEW

**Comment by nonsense** [comment]
I consider implementing `check memory as well, not just CPU`, but we really don't need that now, and this whole functionality will be redundant once we add an autoscaler and/or try to introduce spot instances.

The issue right now is that, if we don't add resource check (which is not race-free, and in general not 100% accurate), developers can try to schedule big workloads, and they would be blocked in `Pending` state on Kubernetes, without much feedback to Testground.

---

This is strictly better than `master`, and I think is good enough for now.

File: pkg/api/composition.go [hunk 1]
```diff
@@ -59,10 +59,18 @@ type Metadata struct {
 	Author string `toml:"author" json:"author"`
 }
 
+type Resources struct {
+	Memory string `toml:"memory"`
+	CPU    string `toml:"cpu"`
```
Comment (raulk) [inline pkg/api/composition.go:64]: Missing json tag.

File: pkg/api/composition.go [hunk 2]
```diff
@@ -59,10 +59,18 @@ type Metadata struct {
...
+}
+
 type Group struct {
 	// ID is the unique ID of this group.
 	ID string `toml:"id" json:"id"`
 
+	// Resources requested for each pod from the Kubernetes cluster
+	Resources Resources `toml:"resources"`
```
Comment (raulk) [inline pkg/api/composition.go:72]: Missing json tag.

**Comment by nonsense** [comment]
> If we add this in the composition schema, we should apply in local:docker also, and emit a warning when set on local:exec. Otherwise, the feature is uneven and it's not clear which runners support it.

By `emit a warning`, you mean emit a warning that these values are ignored, right?

For `local:exec` they would totally be ignored as we have no isolation.

For `local:docker` do we want to apply resource restrictions at all?

**Comment by nonsense** [comment]
@raulk I added a warning for `local:exec` and `local:docker` in case resources are set in the composition file for a given group.

File: pkg/runner/cluster_k8s.go [hunk 3]
```diff
@@ -886,11 +881,28 @@ func (c *ClusterK8sRunner) maxPods(podResourceCPU resource.Quantity) (int, error
...
+
+		neededCPUs += podCPU * float64(g.Instances)
+	}
+
+	if (availableCPUs * utilisation) > neededCPUs {
+		return true, nil
+	}
 
-	return pods, nil
+	ow.Warnw("not enough resources on cluster", "availableCPUs", availableCPUs, "neededCPUs", neededCPUs, "utilisation", utilisation)
```
Comment (raulk) [inline pkg/runner/cluster_k8s.go:904]: nit: lower snake case for logging fields.

**Pull request merged**